src/uct/sm/scopy/base/scopy_ep.c

```diff
@@ -3,21 +3,203 @@
  * See file LICENSE for terms.
  */
 
+#include "scopy_iface.h"
 #include "scopy_ep.h"
 
+#include <uct/base/uct_iov.inl>
+
+
+const char* uct_scopy_tx_op_str[] = {
+    [UCT_SCOPY_TX_PUT_ZCOPY] = "uct_scopy_ep_put_zcopy",
+    [UCT_SCOPY_TX_GET_ZCOPY] = "uct_scopy_ep_get_zcopy"
+};
 
 UCS_CLASS_INIT_FUNC(uct_scopy_ep_t, const uct_ep_params_t *params)
 {
     uct_scopy_iface_t *iface = ucs_derived_of(params->iface, uct_scopy_iface_t);
 
     UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super.super);
 
+    ucs_arbiter_group_init(&self->arb_group);
+
     return UCS_OK;
 }
 
 static UCS_CLASS_CLEANUP_FUNC(uct_scopy_ep_t)
 {
-    /* No op */
+    ucs_arbiter_group_cleanup(&self->arb_group);
 }
 
 UCS_CLASS_DEFINE(uct_scopy_ep_t, uct_base_ep_t)
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_scopy_ep_prepare_tx(uct_ep_h tl_ep, const uct_iov_t *iov,
+                        size_t iov_cnt, uint64_t remote_addr,
+                        uct_rkey_t rkey, uct_completion_t *comp,
+                        uct_scopy_tx_op_t tx_op)
+{
+    uct_scopy_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_scopy_iface_t);
+    uct_scopy_ep_t *ep       = ucs_derived_of(tl_ep, uct_scopy_ep_t);
+    uct_scopy_tx_t *tx;
+    size_t iov_it;
+
+    ucs_assert((tx_op == UCT_SCOPY_TX_PUT_ZCOPY) ||
+               (tx_op == UCT_SCOPY_TX_GET_ZCOPY));
+
+    UCT_CHECK_IOV_SIZE(iov_cnt, iface->config.max_iov, uct_scopy_tx_op_str[tx_op]);
+
+    tx = ucs_mpool_get_inline(&iface->tx_mpool);
+    if (ucs_unlikely(tx == NULL)) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    tx->comp            = comp;
+    tx->rkey            = rkey;
+    tx->remote_addr     = remote_addr;
+    tx->op              = tx_op;
+    tx->consumed_length = 0;
```

Comment:
do we really need to keep consumed_length and total length?
can't we just use iov iterator functions to know when it's finished?